dali/python/nvidia/dali/pipeline.py

```diff
@@ -410,6 +410,9 @@ def _prepare_graph(self, define_graph = None):
             if isinstance(outputs[i], types.ScalarConstant):
                 import nvidia.dali.ops
                 outputs[i] = nvidia.dali.ops._instantiate_constant_node("cpu", outputs[i])
+            elif isinstance(outputs[i], (list, tuple)) and any(isinstance(x, DataNode) for x in outputs[i]):
+                raise RuntimeError("Pipeline output cannot be a list or tuple of DataNodes, only "
+                                   "data types convertible to ``types.Constant`` are allowed.")
```

Comment:
I'd add the following functions (locally or prefixed with `_` private prefix):
```Python
def in_nested(container, pred, container_type):
    if isinstance(container, container_type):
        if any(pred(x) for x in container):
            return True
        for x in container:
            if in_nested(x, pred, container_type):
                return True
    return False
    
def contains_nested_datanode(nested):
    return in_nested(nested, lambda x: isinstance(x, dali.pipeline.DataNode), (list, tuple))
```
also, the error message doesn't say that _unwrapped_ DataNodes are the primary type expected by set outputs. I suggest the following:
```suggestion
            elif contains_nested_datanode(outputs[i]):
                raise TypeError(f"Illegal pipeline output type. The output {i} contains a nested `DataNode`. Allowed types are ``DataNode`` or types from which `types.Constant` can be constructed.")
```

Reply:
Done